llm/main.py

```diff
@@ -58,37 +58,45 @@ def build_scheduler(cfg):
         raise ValueError(f'Not sure how to build scheduler: {cfg.name}')
 
-# Coming soon: this conversion math will be done inside Composer Trainer rather than entrypoint
-def update_batch_size_info(cfg):
-    global_train_batch_size = cfg.global_train_batch_size
-    device_train_batch_size = global_train_batch_size // dist.get_world_size()
-    device_train_microbatch_size = cfg.device_train_microbatch_size
-    if device_train_microbatch_size == 'auto':
-        device_train_grad_accum = 'auto'
-        device_eval_microbatch_size = 'auto'
-        device_eval_batch_size = device_train_batch_size
-    elif isinstance(device_train_microbatch_size, int):
-        if device_train_microbatch_size > device_train_batch_size:
-            print (f"WARNING: device_train_microbatch_size > device_train_batch_size, will be reduced from {device_train_microbatch_size} -> {device_train_batch_size}.")
-            cfg.device_train_microbatch_size = device_train_batch_size
-            device_train_microbatch_size = device_train_batch_size
-        device_train_grad_accum = device_train_batch_size // device_train_microbatch_size
-        device_eval_microbatch_size = device_train_microbatch_size
-        device_eval_batch_size = device_eval_microbatch_size
+def calculate_batch_size_info(global_batch_size, device_microbatch_size):
+    device_batch_size = global_batch_size // dist.get_world_size()
+    device_microbatch_size = device_microbatch_size
+    if device_microbatch_size == 'auto':
+        device_grad_accum = 'auto'
+    elif isinstance(device_microbatch_size, int):
+        if device_microbatch_size > device_batch_size:
+            print (f"WARNING: device_microbatch_size > device_batch_size, will be reduced from {device_microbatch_size} -> {device_batch_size}.")
+            device_microbatch_size = device_batch_size
+            device_microbatch_size = device_batch_size
+        device_grad_accum = device_batch_size // device_microbatch_size
     else:
         raise ValueError(
-            f'Not sure how to parse {device_train_microbatch_size=}')
+            f'Not sure how to parse {device_microbatch_size=}')
+    
+    return device_batch_size, device_grad_accum
+
+
+# Coming soon: this conversion math will be done inside Composer Trainer rather than entrypoint
+def update_batch_size_info(cfg):
+
+    device_train_batch_size, device_train_grad_accum = calculate_batch_size_info(
+        cfg.global_train_batch_size, cfg.device_train_microbatch_size
+    )
+    
+    device_eval_batch_size, _ = calculate_batch_size_info(
+        cfg.global_eval_batch_size, cfg.device_eval_microbatch_size
+    )
 
     cfg.n_gpus = dist.get_world_size()
     cfg.device_train_batch_size = device_train_batch_size
     cfg.device_train_grad_accum = device_train_grad_accum
     cfg.device_eval_batch_size = device_eval_batch_size
-    cfg.device_eval_microbatch_size = device_eval_microbatch_size
+    cfg.device_eval_microbatch_size = device_eval_batch_size
```

Comment:
I'm a bit worried here, I think Composer also has an OOM checker for eval, but if we want to be extra safe, we should probably not try to rely on it, and just use safe config values. For that reason I think we should have the user config include a `device_eval_batch_size` that is set to a safe value, and not worry too much about the global eval batch size, as we're not doing any optimization.

The failure case right now is if `global_eval_batch_size=128` or something is hardcoded, then I can either hit a bug if I try to run on more GPUs, or run into an OOM if I run on too few GPUs.

Reply:
Fixed this in a way that definitely won't break, but is a bit conservative (basically, not relying on Composer's auto eval loop)